### DISCUSSION TITLE
Edge attributes integration and NeighborSampler

### DISCUSSION THREAD

**Discussion by matteomedioli** [pyg-team/pytorch_geometric#2522] [Q&A]
Hi there!
I'm working on a multi-relations graph convolutional task in order to obtain node embeddings of Wordnet. I'm using your implementation of GraphSAGE Unsupervised with NeighborSampler. Since for my downstream task, relations between synset or lemma nodes have a strong influence I was wondering if there's a way to include `data.edge_attrs`(which stores the edge label that identifies the relation type) in SAGEConv (and GCNConv or CATConv too) in the algorithm. 

I have two main doubts (sorry in advance for any inaccuracy):

1. Convolutional Layers such as SAGEConv, GCNConv, and GATConv present `self.propagate()` and `self.message()` in their forward() method. 
From what I understand, conv layers that manage `edge_attrs` (GENConv i.e) present in the `self.propagate()` call the parameter `edge_attr` ([line 152 GENConv](https://github.com/rusty1s/pytorch_geometric/blob/e6b8d6427ad930c6117298006d7eebea0a37ceac/torch_geometric/nn/conv/gen_conv.py#L152)). 
Given these two assumptions, can I simply add the `edge_attrs` in the `forward()` method for example in SAGEConv, and pass it to self.propagate(), `self.message()` in order to integrate multi-relations message passing SAGEConv? ([add here](https://github.com/rusty1s/pytorch_geometric/blob/e6b8d6427ad930c6117298006d7eebea0a37ceac/torch_geometric/nn/conv/sage_conv.py#L69))

2. Since I can only perform unsupervised training over Wordnet, I was wondering to train the model using negative sampling, as you present in `graph_sage_unsup.py` using [this](https://github.com/rusty1s/pytorch_geometric/blob/e6b8d6427ad930c6117298006d7eebea0a37ceac/examples/graph_sage_unsup.py#L22) `NeighborSampler` implementation. There's a way to manage adge_attrs when sampling neg and pos batch?

Many thanks in advance and thanks for sharing work!

**Comment by rusty1s** [comment]
1. It depends. If your `edge_attr` corresponds to an edge type, e.g., holding the relation type of an edge as a `LongTensor`, I suggest to look at the `RGCNConv`, where you apply different `SAGEConv` modules for each distinct edge type. Otherwise, yes, you can simply concatenate or add the `edge_attr` to the neighboring node features in a customized `SAGEConv` implementation.
2. You need to sample or generate random `edge_attr` values as well. If that is just an edge type, you can simply sample any type from the number of relations.
